### ISSUE TITLE
Compatibility with units that do now show indoor temperature

### ISSUE THREAD

**Issue by mstendah** [homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform#31]
Hi, would you be able to make this work more nicely with units that do not support the indoor temperature, as this now both creates and error in the log over an over again, as well as shows 0C as the indoor temerature, which off course is false.

**Comment by embee8** [comment]
Which model do you use? I assume that your indoor unit still has a temperature sensor (so that it knows when to cool/heat) but reports it differently or doesn't expose it via the API?

**Comment by rabbadabba** [comment]
Hi, I've got a CS-Z50TKR Aero Series. (using the CZ-TACG1 wifi controller kit).   Internal unit temperature on the accessory icon is always 0 and the logs say "Temperature State is Unavailable".  Did not have this issue with the previous plugin by codyc1515.  Hope this helps.

**Comment by embee8** [comment]
Please switch on debug mode and share some of your logs here. Specifically the responses from the Comfort Cloud service. This should give us a good idea of what we are missing.

**Comment by 488-NotAcceptableHere** [comment]
Have the same issue (and same combo of unit/wifi controller).  The previous plugin did show the internal unit temperature for these.

Setting to debug though still only logs [Homebridge Panasonic AC Platform] Temperature state is not available

**Comment by embee8** [comment]
Thanks! What would be a desired outcome in your case, i.e. what would you like the Home app to display?

**Comment by 488-NotAcceptableHere** [comment]
That’s an interesting one. Looking at the code for the old plug-in it also reported 0 when it received 126 from Comfort Cloud. 

From a UX perspective I’m not sure what’s better. In the absence of a valid value the options might be continue to report 0, or perhaps report the target temperature? Neither would be accurate though.

**Comment by rabbadabba** [comment]
Did previous versions of the old control do that?   I might have a look at a few other controls and see what they do.

**Comment by embee8** [comment]
>I might have a look at a few other controls and see what they do.

That's great - please report back if you find out more.

>From a UX perspective I’m not sure what’s better. In the absence of a valid value the options might be continue to report 0, or perhaps report the target temperature? Neither would be accurate though.

For transparency and clarity, an option would be to report a special value, like -99. But once you know that your system doesn't report on the indoor temperature, that approach doesn't provide any utility any longer.

With that in mind, the best option might be to display the target temperature, even though that would confuse the status icon of the Home app (it would show 'idle' because it thinks the target temperature has already been reached).

**Comment by rabbadabba** [comment]
FYI in the Cloud Comfort app no temp is reported for internal (has a dash) only external.

**Comment by mstendah** [comment]
Yes, this is correct. According to the documentation, indoor temperature is not supported by all models. Strange, as the indoor unit must measure indoor temperature, but apprently not part of the API/output to Comfort Cloud.
________________________________

**Comment by mstendah** [comment]
Hi, I have a Panasonic CZ-TACG1 WiFi Comfort Cloud remote connected to my Panasonic UZ12-SKE indoor unit.

-mikael

From: embee8 ***@***.***>
Date: Monday, 25. July 2022 at 3.13
To: embee8/homebridge-panasonic-ac-platform ***@***.***>
Cc: mstendah ***@***.***>, Author ***@***.***>
Subject: Re: [embee8/homebridge-panasonic-ac-platform] Compatibility with units that do now show indoor temperature (Issue #31)

Which model do you use? I assume that your indoor unit still has a temperature sensor (so that it knows when to cool/heat) but reports it differently or doesn't expose it via the API?

**Comment by JordyThery** [comment]
I have an CS-Z50UD3EAW and it also does not report an indoor temperature. Would be nice to either disable the reading or have it display something else (e.g the desired temperature). :-)

**Comment by 1homebridge** [comment]
I have this exact problem, opened an issue as well lower down. 
i also get a 0.0*C reading for the exact same reason, that some non commercial units do not report the indoor temperature, so the reading comes up as 0*C. 
In the last update of the previous plug-in for Panasonic AC by Cody, he allowed the outdoor unit which does report the temperature to replace the 0*C value and therefore allowed the units to report Cooling mode properly. As it stands the 0*C reading infers you are in a freezing room, so 16*C wouldn’t activate Cooling mode. 
Hope this helps and can be resolved. 
Thanks

**Comment by mkz212** [comment]
In my AC indoor unit temperature sensor show that is 26 C. Target is 30 C. And real temp in room is 21 C. If i set target temp in AC to 21 it will quickly stop heating. Build-in temperature sensors in AC is mistake! It's like putting a thermometer on the sun.

**Comment by mkz212** [comment]
I think that easy way to fix this is to set default current temp to 16 C in heating mode and 30 in coooing mode. And when indoor unit have temperature sensor it will overide this values.

**Comment by mkz212** [comment]
Maybe in dist/indoor-unit.js add this in Current Temperature:


`this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, (deviceStatus.operationMode === 3) ? 86 : 68,8 );`

This will set default current temperatures if cannot read from indoor unit. 86 Farenheit = 30 Celcius and 68,8 Farenheit = 16 Celcius. I'am not sure that there should be , (comma) or . (dot) in second value. This should set default current temperature to 16 Celcius, and 30 Celcius in heat mode. I don't know how to solve this for auto mode. Set it based on time? - e.g. in the months from May to September, the default cooling mode and outside of these months, heating mode? Or not set anything because if the air conditioning does not have a built-in sensor, the user should not use the auto mode?


and second thing is to replace temp from outdoor unit if user set that in config:

```
 if ((deviceStatus.outTemperature < 126) && (this.platformConfig.useOutdoorTemp === 1) {
                    this.platform.log.error('Outdoor temperature is used instead of indoor');				
					this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.outTemperature);
                }
```




So the whole code would look like this:

```
// Current Temperature
            // Note: Only update the temperature when the heat pump is reporting a valid temperature.
            // Otherwise it will just incorrectly report zero to HomeKit.
            if (deviceStatus.insideTemperature >= 126) {
                // Temperature of 126 from the API = null/failure
                this.platform.log.error('Temperature state is not available - setting default temperature');
		this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, (deviceStatus.operationMode === 3) ? 86 : 68,8 );
				
                if ((deviceStatus.outTemperature < 126) && (this.platformConfig.useOutdoorTemp === 1) {
                      this.platform.log.error('Outdoor temperature is used instead of indoor');				
		      this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.outTemperature);
                }
					
            }
            else {
                this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.insideTemperature);
            }
```

**Comment by mstendah** [comment]
I’m pretty sure the unit has an indoor temp sensor, but it just doesn’t publish that data. You can even see a thermostat sensor under the cover next to the filters, and how could it work without one as the whole point is to set the desired indoor temperature, no matter if you use it for heating or cooling. -mikael   From: mkz212 ***@***.***>Date: Tuesday, 31. January 2023 at 10.39To: embee8/homebridge-panasonic-ac-platform ***@***.***>Cc: mstendah ***@***.***>, Author ***@***.***>Subject: Re: [embee8/homebridge-panasonic-ac-platform] Compatibility with units that do now show indoor temperature (Issue #31)Maybe in dist/indoor-unit.js add this in Current Temperature:this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, (deviceStatus.operationMode === 3) ? 86 : 68,8 );This will set default current temperatures if cannot read from indoor unit. 86 Farenheit = 30 Celcius and 68,8 Farenheit = 16 Celcius. I'am not sure that there should be , (comma) or . (dot) in second value. This should set default current temperature to 16 Celcius, and 30 Celcius in heat mode. I don't know how to solve this for auto mode. Set it based on time? - e.g. in the months from May to September, the default cooling mode and outside of these months, heating mode? Or not set anything because if the air conditioning does not have a built-in sensor, the user should not use the auto mode?and second thing is to replace temp from outdoor unit if user set that in config: if ((deviceStatus.outTemperature < 126) && (this.platformConfig.useOutdoorTemp === 1) {                    this.platform.log.error('Outdoor temperature is used instead of indoor');                                                                    thisservice.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.outTemperature);                }So the whole code would look like this:// Current Temperature            // Note: Only update the temperature when the heat pump is reporting a valid temperature.            // Otherwise it will just incorrectly report zero to HomeKit.            if (deviceStatus.insideTemperature >= 126) {                // Temperature of 126 from the API = null/failure                this.platform.log.error('Temperature state is not available - setting default temperature');                this.service.updateCharacteristic(this.platformCharacteristic.CurrentTemperature, (deviceStatus.operationMode === 3) ? 86 : 68,8 );                                               if ((deviceStatusoutTemperature < 126) && (this.platformConfig.useOutdoorTemp === 1) {                      this.platform.log.error('Outdoor temperature is used instead of indoor');                                                  this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.outTemperature);                }                                                  }            else {                this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.insideTemperature);            }—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

**Comment by mkz212** [comment]
Or better way:

```
if ((deviceStatus.outTemperature < 126) && (this.platformConfig.useOutdoorTemp === 1) {
   this.platform.log.error('Outdoor temperature is used instead of indoor');
   this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.outTemperature);
}
else {
   this.platform.log.error('Temperature state is not available - setting default temperature');		
   this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, (deviceStatus.operationMode === 3) ? 86 : 68,8);
}
```

**Comment by mkz212** [comment]
So whole code:


```
// Current Temperature
// Note: Only update the temperature when the heat pump is reporting a valid temperature.
// Otherwise it will just incorrectly report zero to HomeKit.
         
if (deviceStatus.insideTemperature >= 126) {
// Temperature of 126 from the API = null/failure

  if ((deviceStatus.outTemperature < 126) && (this.platformConfig.useOutdoorTemp === 1) {
   this.platform.log.error('Outdoor temperature is used instead of indoor');
   this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.outTemperature);
  }
  else {
   this.platform.log.error('Temperature state is not available - setting default temperature');		
   this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, (deviceStatus.operationMode === 3) ? 86 : 68,8);
  }				
}
					
else {              
  this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.insideTemperature);
}
```

**Comment by mkz212** [comment]
Another possibility. Fully automatic. If the temperature of the indoor unit is not available, the temperature of the outdoor unit will be used. If both are not available, the default values will be used: 8°C for heating and 30°C for cooling.



```
// Current Temperature
// If the temperature of the indoor unit is not available, the temperature of the outdoor unit will be used. 
// If both are not available, the default values will be used: 8°C for heating and 30°C for cooling.
           

// Temperature of 126 from the API = null/failure
if (deviceStatus.insideTemperature < 126) {
   this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.insideTemperature);
}

else {
   this.platform.log.info('Indoor temperature: is not available');
				
   if (deviceStatus.outTemperature < 126) {
      this.connectedOutdoorUnit.setOutdoorTemperature(deviceStatus.outTemperature);
   }
				
   else {
      this.platform.log.info('Outdoor temperature is not available - setting default temperature');		
      this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, (deviceStatus.operationMode === 3) ? 30 : 8);			  
   }
					
}
```

**Comment by cstromme** [comment]
I just came here to file this same issue. Is there any way to just show "—" instead of "0,0"?

**Comment by mkz212** [comment]
I think that NO. Homekit require that it must be numeric value.

**Comment by cstromme** [comment]
Incredibly annoying. And the "heating to" and "cooling to" are determined by this value as well?

Could we select another temperature sensor in the same room to fill in this value?

**Comment by cstromme** [comment]
That's what our installer said as well. The ones that still comes with one is due to older customers demanding one, but in reality there's little sense in having them on the device because they're too close to the output of the heat pump to be able to accurately read the room temperature.

**Comment by mkz212** [comment]
I already gave the solution a few comments above. You can make changes to the code yourself. It is up to the plugin author to implement these solutions and make them available to everyone.

**Comment by 1homebridge** [comment]
Hi
Thanks for the reply, I read your posts, but I’m not remotely as savvy with code so wasn’t sure which post applied to using the outdoor unit temperature. Could you give me the post number and if not too much trouble, could you let me know of I only this in settings or configuration? I’m showing my lack of knowledge. Thanks again

**Comment by mkz212** [comment]
> I have fully automatic solution. If the temperature of the indoor unit is not available, the temperature of the outdoor unit will be used. If both are not available, the default values will be used: 8°C for heating and 30°C for cooling.
> 
> In dist/indoor-unit.js :
> 
> ```
> // Current Temperature
> // If the temperature of the indoor unit is not available, the temperature of the outdoor unit will be used. 
> // If both are not available, the default values will be used: 8°C for heating and 30°C for cooling.
>            
> 
> // Temperature of 126 from the API = null/failure
> if (deviceStatus.insideTemperature < 126) {
>    this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.insideTemperature);
> }
> 
> else {
>    this.platform.log.info('Indoor temperature: is not available');
> 				
>    if (deviceStatus.outTemperature < 126) {
>       this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, deviceStatus.outTemperature);
>    }
> 				
>    else {
>       this.platform.log.info('Outdoor temperature is not available - setting default temperature');		
>       this.service.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, (deviceStatus.operationMode === 3) ? 30 : 8);			  
>    }
> 					
> }
> ```

**Comment by 1homebridge** [comment]
Thanks again,
I just don’t know hope to input this to the json editor. Is it in the config?
Sorry

**Comment by mkz212** [comment]
No. You need to ftp your device to explore plugin folder. You can search on net how to do it.

**Comment by 1homebridge** [comment]
Thanks again, I’m currently viewing on an iPad and can’t see where to edit the code and don’t want to do something wrong. Don’t want to test your goodwill tho. I was hoping embee8 could update this, mentioned I was willing to sponsor the plug-in. If anyone can guide me through this process, I’d very much appreciate it.  

Thanks

**Comment by cstromme** [comment]
Any chance we could get a setting to let it know we don't have a outdoor temperature sensor? So that it stops spamming the logs with "Temperature state is not available"?

**Comment by mkz212** [comment]
Dodałem pull request w tej sprawie #87

**Comment by mkz212** [comment]
I added a pull request on this #87

**Comment by jandersonhill** [comment]
This feature has been made available from v1.8.0 onwards
